Samples/StyleTransfer/VideoEffect/StyleTransferEffectCpp/StyleTransferEffect.cpp

```diff
@@ -0,0 +1,150 @@
+﻿#include "pch.h"
+#include "StyleTransferEffect.h"
+#include "StyleTransferEffect.g.cpp"
+#include <ppltasks.h>
+#include <sstream>
+
+using namespace std;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::Storage::Streams;
+using namespace concurrency;
+
+namespace winrt::StyleTransferEffectCpp::implementation
+{
+	StyleTransferEffect::StyleTransferEffect() :
+		Session(nullptr)
+	{
+		for (int i = 0; i < swapChainEntryCount; i++)
+		{
+			bindings.push_back(std::make_unique<SwapChainEntry>());
+		}
+	}
+
+	IVectorView<VideoEncodingProperties> StyleTransferEffect::SupportedEncodingProperties() {
+		VideoEncodingProperties props = VideoEncodingProperties();
+		props.Subtype(L"ARGB32");
+		return single_threaded_vector(std::vector<VideoEncodingProperties>{props}).GetView();
+	}
+
+	bool StyleTransferEffect::TimeIndependent() { return true; }
+	MediaMemoryTypes StyleTransferEffect::SupportedMemoryTypes() { return MediaMemoryTypes::GpuAndCpu; }
+	bool StyleTransferEffect::IsReadOnly() { return false; }
+	void StyleTransferEffect::DiscardQueuedFrames() {}
+
+	void StyleTransferEffect::Close(MediaEffectClosedReason) {
+		OutputDebugString(L"Close Begin | ");
+		std::lock_guard<mutex> guard{ Processing };
+		// Make sure evalAsyncs are done before clearing resources
+		for (int i = 0; i < swapChainEntryCount; i++) {
+			if (bindings[i]->activetask != nullptr &&
+				bindings[i]->binding != nullptr)
+			{
+				std::wostringstream ss;
+				ss << i;
+				OutputDebugString(ss.str().c_str());
+				bindings[i]->activetask.get();
+				bindings[i]->binding.Clear();
+			}
+		}
+		if (Session != nullptr) Session.Close();
+		OutputDebugString(L"Close\n");
+	}
+
+	void StyleTransferEffect::SubmitEval(VideoFrame input, VideoFrame output) {
+		auto currentBinding = bindings[0].get();
+		if (currentBinding->activetask == nullptr
+			|| currentBinding->activetask.Status() != Windows::Foundation::AsyncStatus::Started)
```

Comment:
You need some type of lock or atomic access everywhere where variables are shared between threads, not only on write. `bindings[0]` might get moved by the `std::rotate(bindings.begin(), bindings.begin() + 1, bindings.end());` below.

Reply:
Added lock guard to setting currentBinding.

Reply:
AFAIK this is not needed.

bindings is only touched by a single thread (with the exception of calling set properties, which is not supported while things are running.

the eval async completed handler has a pointer to the underlying memory and not the bindings object, so things shold be fine.

---
In reply to: [465236562](https://github.com/microsoft/Windows-Machine-Learning/pull/335#discussion_r465236562) [](ancestors = 465236562)